### PR TITLE
Added auto-connect feature for easy remote work on startup and "multiple" connections

### PR DESCRIPTION
**Pull request by andymule**
Allows users to save without explicitly connecting to a server. Attempts to match profile names with username parsed from filepath, and if a match is found a connection is opened and the file is uploaded. This enables users to instantly resume working remotely without considering if they're saving files on the server or not.

### CODE REVIEW

**Comment by andymule** [comment]
9dc0b3f might go too far for this initial scope, as it tries to automatically connect to the necessary host for whatever file is saved. but the code is almost identical for the original scope so i think we should go for it

**Comment by andymule** [comment]
bug: needs to NOT disconnect on local files

File: src/NppFTP.cpp [hunk 1]
```diff
@@ -305,57 +305,55 @@ void NppFTP::AttemptToAutoConnect(const TCHAR* path)
...
-		TCHAR username[FILENAME_MAX];	// assume username+path is smaller than largest allowed filename
-		memset(username, '\0', FILENAME_MAX);
-		int nameLength = (atSymbolIndex - previousBackslashIndex - 1);
+		TCHAR username[FILENAME_MAX*2];	// assume username+path is smaller than largest allowed filename, x2 because TCHAR is 2 bytes
+		TCHAR hostname[FILENAME_MAX * 2];	// assume username+path is smaller than largest allowed filename, x2 because TCHAR is 2 bytes
+		memset(username, '\0', FILENAME_MAX*2);
+		memset(hostname, '\0', FILENAME_MAX * 2);
+		int nameLength = (atSymbolIndex - previousBackslashIndex);
+		int hostnameLength = (nextBackslashIndex - atSymbolIndex - 1);
 		memcpy(username, &path[previousBackslashIndex], nameLength * 2);	// TCHAR is 2 bytes so we have to double the nameLength to be correct in byte-space
```
Comment (chcg) [inline src/NppFTP.cpp]: use sizeof(TCHAR) instead of hardcoded 2 (for old ascii builds TCHAR would be just 1 byte)

Comment (andymule) [reply src/NppFTP.cpp]: nice catch

**Comment by andymule** [comment]
@ashkulz Can I get a review on this? Interested in your thoughts.

**Comment by andymule** [comment]
DLL for testing if interested:
https://drive.google.com/file/d/1KfQiVjd1PPl_Ic7Xpkiv_kW699ScoJxi/view?usp=sharing

Just drop into your plugins folder, mine is here:
C:\Program Files (x86)\Notepad++\plugins\

**Comment by chcg** [comment]
You could also find a build at appveyor for some time:
x64 https://ci.appveyor.com/project/ashkulz/nppftp/build/0.27.4.58/job/17uho6e0bv7ttg6y/artifacts
x32 https://ci.appveyor.com/project/ashkulz/nppftp/build/0.27.4.58/job/yu316rudr8tcchrj/artifacts

@andymule I'm quite sure ashkulz won't step in to have a look at this PR. I could do the merge, but most probably that will not take place before September. So please be patient with me.

**Comment by andymule** [comment]
Awesome, sounds great. Timeframe seems fine. Would love to get more people testing it out, might blast it onto some related forum posts....

**Comment by andymule** [comment]
bug: If you quickly save two files on different servers very quickly, it's possible that the first file will be disconnected mid-save

**Comment by ashkulz** [comment]
I haven't used Windows for 2+ years, so @chcg is the actual maintainer.

> @andymule <https://github.com/andymule> I'm quite sure ashkulz won't step
> in to have a look at this PR. I could do the merge, but most probably that
> will not take place before September. So please be patient with me.
>

**Comment by chcg** [comment]
@andymule In general the PR looks fine. I think what is still needed is a general option to enable/disable this behaviour to allow the user to decide what he wants. What about the last bug you saw on the two quick saves. "Quick" in this context seems to be related to the connection speed to the server.

**Pull request merged**